File: 0113-path-sum-ii/0113-path-sum-ii.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} targetSum
 * @return {number[][]}
 */
var pathSum = function(root, targetSum) {
    const pathSumEqualsTargetAllPaths = [];

    function traverse(currentNode, currPathSum, currPathNodesArr){
        if(currentNode === null){
            return;
        }
        const currentNodeVal = currentNode.val;

        currPathSum += currentNodeVal;

        currPathNodesArr.push(currentNodeVal);

        traverse(currentNode.left, currPathSum, [...currPathNodesArr]);
        traverse(currentNode.right, currPathSum, [...currPathNodesArr]);

        // Leaf node condition;
        if(currentNode.left === null && currentNode.right === null){
            const isPathSumEqual = currPathSum === targetSum;
            if(isPathSumEqual){
                pathSumEqualsTargetAllPaths.push(currPathNodesArr);
            }
        }
    }

    traverse(root, 0, []);

    return pathSumEqualsTargetAllPaths;
};

File: 0113-path-sum-ii/0113-path-sum-ii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        pathsum_equals_target_all_paths = []

        def traverse(current_node, curr_pathsum, curr_pathnodes_arr):
            if not current_node:
                return

            current_node_val = current_node.val
            curr_pathsum += current_node_val

            curr_pathnodes_arr.append(current_node_val)        

            traverse(current_node.left, curr_pathsum, curr_pathnodes_arr.copy())
            traverse(current_node.right, curr_pathsum, curr_pathnodes_arr.copy())

            # leaf node condition
            if current_node.left == None and current_node.right == None:
                is_pathsum_equal = curr_pathsum == targetSum

                if is_pathsum_equal:
                    pathsum_equals_target_all_paths.append(curr_pathnodes_arr)
                    
        traverse(root, 0, [])

        return pathsum_equals_target_all_paths